frontend/apps/reader/modules/readerhighlight.lua

```diff
@@ -1751,133 +1731,41 @@ function ReaderHighlight:highlightFromHoldPos()
-        }
-    end
-end
-
-function ReaderHighlight:saveHighlight(extend_to_sentence)
-    self.ui:handleEvent(Event:new("AddHighlight"))
-    logger.dbg("save highlight")
-    if self.selected_text and self.selected_text.pos0 and self.selected_text.pos1 then
-        if extend_to_sentence and self.ui.rolling then
-            local extended_text = self.ui.document:extendXPointersToSentenceSegment(self.selected_text.pos0, self.selected_text.pos1)
-            if extended_text then
-                self.selected_text = extended_text
+        self.ui:handleEvent(Event:new("AddHighlight"))
+        local pg_or_xp, page
+        if self.ui.rolling then
+            if extend_to_sentence then
+                local extended_text = self.ui.document:extendXPointersToSentenceSegment(self.selected_text.pos0, self.selected_text.pos1)
+                if extended_text then
+                    self.selected_text = extended_text
+                end
             end
+            pg_or_xp = self.selected_text.pos0
+            page = self.ui.document:getPageFromXPointer(pg_or_xp)
+        else
+            pg_or_xp = self.selected_text.pos0.page
+            page = pg_or_xp
         end
-        local page = self.ui.paging and self.selected_text.pos0.page or self.ui.document:getPageFromXPointer(self.selected_text.pos0)
-        if not self.view.highlight.saved[page] then
-            self.view.highlight.saved[page] = {}
-        end
-        local datetime = os.date("%Y-%m-%d %H:%M:%S")
-        local pg_or_xp = self.ui.paging and page or self.selected_text.pos0
-        local chapter_name = self.ui.toc:getTocTitleByPage(pg_or_xp)
-        local hl_item = {
-            datetime = datetime,
-            text = cleanupSelectedText(self.selected_text.text),
+        local item = {
+            page = self.ui.paging and self.selected_text.pos0.page or self.selected_text.pos0,
             pos0 = self.selected_text.pos0,
             pos1 = self.selected_text.pos1,
-            pboxes = self.selected_text.pboxes,
-            ext = self.selected_text.ext,
+            text = cleanupSelectedText(self.selected_text.text),
             drawer = self.view.highlight.saved_drawer,
-            chapter = chapter_name,
+            chapter = self.ui.toc:getTocTitleByPage(pg_or_xp),
         }
-        table.insert(self.view.highlight.saved[page], hl_item)
-        local bookmark_item = self:getHighlightBookmarkItem()
-        if bookmark_item then
-            bookmark_item.datetime = datetime
-            bookmark_item.chapter = chapter_name
-            self.ui.bookmark:addBookmark(bookmark_item)
+        if self.ui.paging then
+            item.pboxes = self.selected_text.pboxes
+            item.ext = self.selected_text.ext
+            self:writePdfAnnotation("save", page, item)
         end
-        self:writePdfAnnotation("save", page, hl_item)
-        return page, #self.view.highlight.saved[page]
+        return self.ui.bookmark:addItem(item) -- index
```

Comment:
Just asking: so, ReaderHighlight does not update `annotations` directly, but delegates that to ReaderBookmark?